### DISCUSSION TITLE
wdyk_3phase_energymonitor <> 3wjt_3_phase_circuit_breaker

### DISCUSSION THREAD

**Discussion by geedsen** [make-all/tuya-local#2211] [Q&A]
I included two of these 
![image](https://github.com/user-attachments/assets/0a875852-1278-4785-b055-2a142c7fb9fb)
Tuya local recognizes them as wdyk_3phase_energymonitor and a lot of entities are correct. However there are plenty more DP's and some of the same once have slightly different/better specifications.
So I though I would make a copy of the wdyk_3phase_energymonitor.yaml, give it a new name and start adapting it. But adding the device does not give me the possibility to select it.
So I though I would just modify the existing wdyk_3phase_energymonitor.yaml , and when I add it I will get my modified entities. Again no luck, the device is no longer recognized as such. So I dont understand how I should approach this.
I really would appreciate some help how to address this. 
Some of the changes I wanted to make:

```
 - id: 26
      type: bitfield
      name: fault_code
```

change it to:

```
- id: 26
      type: bitfield
      name: fault_code
      mapping:
        - dps_val: 2
          value: overvoltage
        - dps_val: 4
          value: undervoltage
        - dps_val: 8
          value: overcurrent
        - dps_val: 16
          value: temperature
        - dps_val: 32
          value: leakage
        - dps_val: 64
          value: leakagefault
        - dps_val: 128
          value: zerocharge
        - dps_val: 512
          value: maintenance
```

And I wanted to add this one:

```
- id: 109
        name: temperature_threshold
        type: base64
        optional: false
        format:
          - name: temperature
            bytes: 1
            range:
              min: -40
              max: 127
          - name: break
            bytes: 1
            range:
              min: 0
              max: 1
          - name: alarm
            bytes: 1
            range:
              min: 0
              max: 
```

If these are actually 100% correct I dont know. I have more DP's I woould like to add.

This is the model specification:

```
{
    "modelId": "0000003wjt",
    "services": [{
            "actions": [],
            "code": "",
            "description": "",
            "events": [],
            "name": "default_service",
            "properties": [{
                    "abilityId": 1,
                    "accessMode": "rw",
                    "code": "switch_1",
                    "description": "",
                    "extensions": {
                        "iconName": "icon-dp_power2",
                        "attribute": "1"
                    },
                    "name": "switch_1",
                    "typeSpec": {
                        "type": "bool"
                    }
                }, {
                    "abilityId": 9,
                    "accessMode": "rw",
                    "code": "countdown_1",
                    "description": "",
                    "extensions": {
                        "iconName": "icon-dp_time2"
                    },
                    "name": "switch_1_countdown",
                    "typeSpec": {
                        "type": "value",
                        "max": 86400,
                        "min": 0,
                        "scale": 0,
                        "step": 1,
                        "unit": "s"
                    }
                }, {
                    "abilityId": 26,
                    "accessMode": "ro",
                    "code": "fault",
                    "description": "Overvoltage, undervoltage, overcurrent, excessive temperature, excessive leakage current, leakage self-check failure, remaining power is zero, maintenance lock",
                    "extensions": {
                        "iconName": "icon-baojing",
                        "scope": "fault"
                    },
                    "name": "Fault alarm",
                    "typeSpec": {
                        "type": "bitmap",
                        "label": ["OverVoltage", "UnderVoltage", "OverCurrent", "Temperature", "leakage", "LeakageFault", "ZeroCharge", "Maintenance"],
                        "maxlen": 8
                    }
                }, {
                    "abilityId": 38,
                    "accessMode": "rw",
                    "code": "relay_status",
                    "description": "0 means power-on opening, 1 means power-on closing, 2 means power-on hold.",
                    "extensions": {
                        "iconName": "icon-zhuangtai"
                    },
                    "name": "Power-on status settings",
                    "typeSpec": {
                        "type": "enum",
                        "range": ["0", "1", "2"]
                    }
                }, {
                    "abilityId": 40,
                    "accessMode": "rw",
                    "code": "child_lock",
                    "description": "",
                    "extensions": {
                        "iconName": "icon-dp_lock"
                    },
                    "name": "Child lock switch",
                    "typeSpec": {
                        "type": "bool"
                    }
                }, {
                    "abilityId": 101,
                    "accessMode": "ro",
                    "code": "Voltage",
                    "description": "A, B, C voltage, each voltage unit is 0.1V, 2-byte hex value, a total of 6 bytes n high byte first, for example 08 80 08 81 08 82 means A phase 217.6V, B phase 217.7VnC phase 217.8 V",
                    "name": "Voltage",
                    "typeSpec": {
                        "type": "raw",
                        "maxlen": 128
                    }
                }, {
                    "abilityId": 102,
                    "accessMode": "ro",
                    "code": "Current",
                    "description": "Each phase current of A, B, C, 3 bytes per phase, hex, unit 0.001An high byte first, for example 00 03 E8 00 03 E9 00 03 EA means A phase 1.000A, B phase n1.001A, C phase 1.002A",
                    "name": "Current",
                    "typeSpec": {
                        "type": "raw",
                        "maxlen": 128
                    }
                }, {
                    "abilityId": 103,
                    "accessMode": "ro",
                    "code": "ActivePower",
                    "description": "Total active power of 4 channels A, B and C, each channel is 3 bytes, hex, unit 0.0001kWn example 00 27 10 00 27 11 00 27 12 00 27 13 n represents the total power 1.0000kW, A phase power 1.0001kW, B phase power 1.0002kW, C phase power 1.0003kW",
                    "name": "ActivePower",
                    "typeSpec": {
                        "type": "raw",
                        "maxlen": 128
                    }
                }, {
                    "abilityId": 104,
                    "accessMode": "ro",
                    "code": "LeakageCurrent",
                    "description": "Leakage current (0~3000)",
                    "name": "LeakageCurrent",
                    "typeSpec": {
                        "type": "value",
                        "max": 3000,
                        "min": 0,
                        "scale": 0,
                        "step": 1,
                        "unit": "mA"
                    }
                }, {
                    "abilityId": 105,
                    "accessMode": "ro",
                    "code": "Temperature",
                    "description": "1 channel in total, 1 byte per channel, value -40~127 degrees",
                    "name": "Temperature",
                    "typeSpec": {
                        "type": "value",
                        "max": 127,
                        "min": -40,
                        "scale": 0,
                        "step": 1,
                        "unit": "C"
                    }
                }, {
                    "abilityId": 107,
                    "accessMode": "ro",
                    "code": "RemainingEnergy",
                    "description": "Remaining power",
                    "name": "RemainingEnergy",
                    "typeSpec": {
                        "type": "value",
                        "max": 100000000,
                        "min": 0,
                        "scale": 2,
                        "step": 1,
                        "unit": "kWh"
                    }
                }, {
                    "abilityId": 108,
                    "accessMode": "wr",
                    "code": "RechargeEnergy",
                    "description": "For charging capacity",
                    "name": "RechargeEnergy",
                    "typeSpec": {
                        "type": "value",
                        "max": 1000000,
                        "min": 0,
                        "scale": 2,
                        "step": 1,
                        "unit": "kWh"
                    }
                }, {
                    "abilityId": 109,
                    "accessMode": "rw",
                    "code": "CostParameters",
                    "description": "The unit price of electricity bill is 0.0001 yuan, 2 bytes hex, the remaining power is zero trip enable 1 byte, the value is 1 to enable, the others are not enabled n Example 27 10 01 The unit price of electricity is 1.0000 yuan, the remaining power is zero trip enable. nUnit price range of electricity bill: 0.0000 yuan~6.5535 yuan",
                    "name": "CostParameters",
                    "typeSpec": {
                        "type": "raw",
                        "maxlen": 128
                    }
                }, {
                    "abilityId": 110,
                    "accessMode": "rw",
                    "code": "LeakageParameters",
                    "description": "1 byte leakage timing self-test day (hex value 1~28,) 1 byte leakage timing self-test hour (hex value, 0~23) n leakage timing self-test enable (1 byte value is 1 enable, others Disabled), n leakage threshold (2 bytes, hex, unit mA), n leakage over-limit opening enable (1 byte, value is 1 to enable, others are not enabled), leakage over-limit alarm enable Enable (1 byte, the value is 1 to enable, others are not enabled), n1 byte leakage self-check immediately (click once, the value is 1) n example 02 01 01 1E 00 01 01 01 means 1 o'clock on the 2nd of each month Timing leakage self-test, scheduled leakage self-test enablement, leakage threshold 30mA, leakage over-opening enable, leakage over-limit alarm enablement, immediate leakage self-test once n leakage threshold range 1mA~300mA",
                    "name": "LeakageParameters",
                    "typeSpec": {
                        "type": "raw",
                        "maxlen": 128
                    }
                }, {
                    "abilityId": 111,
                    "accessMode": "rw",
                    "code": "VoltageThreshold",
                    "description": "Overvoltage threshold (4 bytes in total, 2 bytes voltage value, unit 0.1V, 1 byte opening enable, value is 1 to enable, other values are not enabled, 1 byte alarm enable, value is 1 to enable Yes, other values are not enabled) n undervoltage threshold (a total of 4 bytes, 2 bytes of voltage value, unit 0.1V, 1 byte of opening enable, the value is 1 to enable, other values ​​are not enabled, 1 word Section alarm is enabled, the value is 1 to enable, other values ​​​​are not enabled) n example 08 98 01 01 08 98 01 01 indicates the overvoltage threshold 220.0v, overvoltage opening enable, overvoltage alarm enable, undervoltage threshold 220.0v, undervoltage trip enable, undervoltage alarm enable n overvoltage threshold range 50V~380V (actual value 500~3800) n undervoltage threshold range 0V~380V (actual value 0~3800)",
                    "name": "VoltageThreshold",
                    "typeSpec": {
                        "type": "raw",
                        "maxlen": 128
                    }
                }, {
                    "abilityId": 112,
                    "accessMode": "rw",
                    "code": "CurrentThreshold",
                    "description": "A total of 5 bytes, 3 bytes of current value, hex, unit 0.001A; 1 byte of opening enable, the value is 1 to enable, other values are not enabled; 1 byte of alarm enable, the value is 1 to enable, Other values are not enabled, n example is 00 07 D0 01 01 overcurrent threshold 2.000A, overcurrent trip enable, overcurrent alarm enable. nCurrent threshold range 0~999A (actual value 0~999000)",
                    "name": "CurrentThreshold",
                    "typeSpec": {
                        "type": "raw",
                        "maxlen": 128
                    }
                }, {
                    "abilityId": 113,
                    "accessMode": "rw",
                    "code": "TemperatureThreshold",
                    "description": "A total of 3 bytes n1 byte temperature threshold (range -40~127); n1 byte opening enable, the value is 1 to enable, other values are not enabled; n1 byte alarm enable, the value is 1 to enable, Other values are not enabled. nExample 30 01 01 Temperature threshold 48 degrees, high temperature opening enabled, high temperature alarm enabled",
                    "name": "TemperatureThreshold",
                    "typeSpec": {
                        "type": "raw",
                        "maxlen": 128
                    }
                }, {
                    "abilityId": 114,
                    "accessMode": "ro",
                    "code": "KWH",
                    "description": "Total active power",
                    "name": "Total active power",
                    "typeSpec": {
                        "type": "value",
                        "max": 100000000,
                        "min": 0,
                        "scale": 2,
                        "step": 1,
                        "unit": "kWh"
                    }
                }, {
                    "abilityId": 115,
                    "accessMode": "ro",
                    "code": "NumberAndType",
                    "description": "Up to 20 displayable ASCII characters n Example MCB2P123456789011n represents model MCB2P device number 123456789011",
                    "name": "NumberAndType",
                    "typeSpec": {
                        "type": "string",
                        "maxlen": 255
                    }
                }, {
                    "abilityId": 116,
                    "accessMode": "wr",
                    "code": "ClearEnergy",
                    "description": "Clear battery",
                    "extensions": {
                        "iconName": "icon-Ele"
                    },
                    "name": "ClearEnergy",
                    "typeSpec": {
                        "type": "bool"
                    }
                }, {
                    "abilityId": 117,
                    "accessMode": "rw",
                    "code": "locking",
                    "description": "If the equipment trips due to leakage, short circuit, etc., the remote device should prompt the operator to unlock it before closing to prevent potential safety hazards.",
                    "extensions": {
                        "iconName": "icon-tcl_function_hs"
                    },
                    "name": "device lock unlock",
                    "typeSpec": {
                        "type": "bool"
                    }
                }, {
                    "abilityId": 118,
                    "accessMode": "ro",
                    "code": "RKWH",
                    "description": "Total reverse active power, range 0~1000000.00kWh",
                    "name": "Total reverse active power",
                    "typeSpec": {
                        "type": "value",
                        "max": 100000000,
                        "min": 0,
                        "scale": 2,
                        "step": 1,
                        "unit": "kWh"
                    }
                }, {
                    "abilityId": 119,
                    "accessMode": "ro",
                    "code": "VRecording",
                    "description": "A, B, C voltage, each voltage unit is 0.1V, 2-byte hex value, a total of 6 bytes n high byte first, for example 08 80 08 81 08 82 means A phase 217.6V, B phase 217.7VnC phase 217.8 V",
                    "name": "Historical voltage",
                    "typeSpec": {
                        "type": "raw",
                        "maxlen": 128
                    }
                }, {
                    "abilityId": 120,
                    "accessMode": "ro",
                    "code": "IRecording",
                    "description": "Each phase current of A, B, C, 3 bytes per phase, hex, unit 0.001An high byte first, for example 00 03 E8 00 03 E9 00 03 EA means A phase 1.000A, B phase n1.001A, C phase 1.002A",
                    "name": "historical current",
                    "typeSpec": {
                        "type": "raw",
                        "maxlen": 128
                    }
                }
            ]
        }
    ]
}

```

And here a device query so you can see some actual values returned:

```
{
  "result": {
    "properties": [
      {
        "code": "switch_1",
        "custom_name": "",
        "dp_id": 1,
        "time": 1723727863617,
        "type": "bool",
        "value": true
      },
      {
        "code": "countdown_1",
        "custom_name": "",
        "dp_id": 9,
        "time": 1723554602233,
        "type": "value",
        "value": 0
      },
      {
        "code": "fault",
        "custom_name": "",
        "dp_id": 26,
        "time": 1723714907673,
        "type": "bitmap",
        "value": 0
      },
      {
        "code": "relay_status",
        "custom_name": "",
        "dp_id": 38,
        "time": 1723554602753,
        "type": "enum",
        "value": "2"
      },
      {
        "code": "child_lock",
        "custom_name": "",
        "dp_id": 40,
        "time": 1723554603151,
        "type": "bool",
        "value": false
      },
      {
        "code": "Voltage",
        "custom_name": "",
        "dp_id": 101,
        "time": 1723744382111,
        "type": "raw",
        "value": "CNsIzgjT"
      },
      {
        "code": "Current",
        "custom_name": "",
        "dp_id": 102,
        "time": 1723744382509,
        "type": "raw",
        "value": "AAfVAAgtAAgM"
      },
      {
        "code": "ActivePower",
        "custom_name": "",
        "dp_id": 103,
        "time": 1723744382528,
        "type": "raw",
        "value": "ACuQAA44AA5aAA79"
      },
      {
        "code": "LeakageCurrent",
        "custom_name": "",
        "dp_id": 104,
        "time": 1723744292831,
        "type": "value",
        "value": 19
      },
      {
        "code": "Temperature",
        "custom_name": "",
        "dp_id": 105,
        "time": 1723743602293,
        "type": "value",
        "value": 43
      },
      {
        "code": "RemainingEnergy",
        "custom_name": "",
        "dp_id": 107,
        "time": 1723554604347,
        "type": "value",
        "value": 0
      },
      {
        "code": "RechargeEnergy",
        "custom_name": "",
        "dp_id": 108,
        "time": 1723554600060,
        "type": "value",
        "value": 0
      },
      {
        "code": "CostParameters",
        "custom_name": "",
        "dp_id": 109,
        "time": 1723741531120,
        "type": "raw",
        "value": "AAAA"
      },
      {
        "code": "LeakageParameters",
        "custom_name": "",
        "dp_id": 110,
        "time": 1723741546139,
        "type": "raw",
        "value": "AQAAABkAAAA="
      },
      {
        "code": "VoltageThreshold",
        "custom_name": "",
        "dp_id": 111,
        "time": 1723741561160,
        "type": "raw",
        "value": "CyIAAAH0AAA="
      },
      {
        "code": "CurrentThreshold",
        "custom_name": "",
        "dp_id": 112,
        "time": 1723741576132,
        "type": "raw",
        "value": "AJCIAQE="
      },
      {
        "code": "TemperatureThreshold",
        "custom_name": "",
        "dp_id": 113,
        "time": 1723741591157,
        "type": "raw",
        "value": "MQEA"
      },
      {
        "code": "KWH",
        "custom_name": "",
        "dp_id": 114,
        "time": 1723744368287,
        "type": "value",
        "value": 15982
      },
      {
        "code": "NumberAndType",
        "custom_name": "",
        "dp_id": 115,
        "time": 1723554605798,
        "type": "string",
        "value": "463100000006        "
      },
      {
        "code": "ClearEnergy",
        "custom_name": "",
        "dp_id": 116,
        "time": 1723554600060,
        "type": "bool",
        "value": false
      },
      {
        "code": "locking",
        "custom_name": "",
        "dp_id": 117,
        "time": 1723554605284,
        "type": "bool",
        "value": false
      },
      {
        "code": "RKWH",
        "custom_name": "",
        "dp_id": 118,
        "time": 1723744368687,
        "type": "value",
        "value": 15976
      },
      {
        "code": "VRecording",
        "custom_name": "",
        "dp_id": 119,
        "time": 1723741201661,
        "type": "raw",
        "value": "CPQI6wjx"
      },
      {
        "code": "IRecording",
        "custom_name": "",
        "dp_id": 120,
        "time": 1723741201680,
        "type": "raw",
        "value": "AA20AA32AA4T"
      }
    ]
  },
  "success": true,
  "t": 1723744383644,
  "tid": "382a1aa65b2f11ef883136d89ae34f8e"
}

```

**Comment by make-all** [comment]
From your description, either one of the following seems to be happening:

1. The yaml you are creating is invalid, and causing the matching to fail
2. The new dps you are adding are not actually present in the device data, causing the matching to fail

**Reply by make-all** [reply]
Reviewing your yaml:
- The indentation is incorrect, entries under `id` should be indented to the same level as the `id`.
- `format` only works for specific cases where the values are processed together (decoding rgb lightbulbs).  To split a dp into separate values, you need to use multiple entries with the same id, different name, and `mask`.

**Reply by geedsen** [reply]
Thanks Jason,  I have another try. Does a (correct) change in the yaml automatically get reflected in the entities when HA restarts?

**Reply by make-all** [reply]
Yes, when HA restarts all the yaml files will be reloaded

**Reply by geedsen** [reply]
So if I look at DP 113, I can see the value being "ZAEA", (I set the temp value to 100)
If in C# (one of the languages I do know :) I can do the following:

```
using System;
					
public class Program
{
	public static void Main()
	{
	string toEncode = "ZAEA";	
    Console.WriteLine(toEncode);
    byte[] regularBytes = Convert.FromBase64String(toEncode);
    foreach(var x in regularBytes)
	   {
	   	  Console.WriteLine(Convert.ToInt32(x));
	   }
	}

}

````

And it displays:


```
ZAEA
100
1
0

````

So those bytes represent the values I see in the app, but I dont understand how I can use the mask to get an individual value from it.
Could you help me here?

**Reply by make-all** [reply]
These mask values will get each of the bytes:
```
mask: ff0000
mask: 00ff00
mask: 0000ff
```

**Reply by geedsen** [reply]
Thanks. I think I tried that, let me try again.

**Reply by geedsen** [reply]
I am not getting anything useful out of it. The entities show, but all with just 'Unavailable'.
```
  - entity: sensor
    name: Temperature Threshold Break
    class: temperature
    category: config
    dps:
      - id: 113
        type: base64
        name: sensor
        class: 
        optional: true
        force: true
        mapping:
          - scale: 1
            mask: "00FF00"

```

**Comment by make-all** [comment]
If that is from the device query above, it is cloud values. It is quite likely that this attribute is not included in the local data, or sent infrequently.
